executor/src/witgen/jit/symbolic_expression.rs

```diff
@@ -0,0 +1,319 @@
+use std::{
+    fmt::{self, Display, Formatter},
+    ops::{Add, BitAnd, BitOr, Mul, Neg},
+    rc::Rc,
+};
+
+use powdr_number::FieldElement;
+
+use crate::witgen::range_constraints::RangeConstraint;
+
+/// A value that is known at run-time, defined through a complex expression
+/// involving known cells or variables and compile-time constants.
+/// Each of the sub-expressions can have its own range constraint.
+#[derive(Debug, Clone)]
+pub enum SymbolicExpression<T: FieldElement, V> {
+    /// A concrete constant value known at compile time.
+    Concrete(T),
+    /// A symbolic value known at run-time, referencing either a cell or a local variable.
+    Variable(V, Option<RangeConstraint<T>>),
```

Comment:
Maybe we call it something else? Like Lucas, I did take a while to understand why coefficients in `AffineSymbolicExpression` can include variables. Maybe `RuntimeValue`?